content/asm_3.md

```diff
@@ -11,243 +14,609 @@ global _start
+Note that although we expect to get two command-line arguments, we are comparing the actual number with `3`. This is because the first implicit argument for every program is its name.
 
-```assembly
-mov rax, r10
-;; number counter
-xor r12, r12
-;; convert to string
-jmp int_to_str
-```
+After making sure that the required number of command-line arguments are passed to our program, we can start working with them. But what do we need to do? Here are the steps:
+
+1. Convert the given command line arguments to integer numbers and calculate the sum of the given numbers.
+2. Convert the result back to string and print it on the screen.
+
+In the next two sections, we will see a detailed explanation of these steps.
+
+### Converting a string to an integer
 
-Here we put sum of command line arguments to rax register, set r12 to zero and jump to int_to_str. Ok now we have base of our program. We already know how to print string and we have what to print. Let's see at str_to_int and int_to_str implementation.
+As the command-line arguments of each program are represented as strings, first we need to convert our command-line arguments to numbers to calculate their sum. To convert a given string to a number, we will use a simple algorithm:
+
+1. Create an accumulator to store an intermediate result while converting the string into its numeric representation.
+2. Take the first byte of the string and subtract the value `48` from it. Each byte in a string is an [ASCII](https://en.wikipedia.org/wiki/ASCII) symbol with its own code. The symbol 0 has code `48`, the symbol 1 has code `49`, and so on. If we subtract `48` from the ASCII code of the given symbol, we get an integer representation of the current digit from the given string.
+3. As soon as we know the current digit, we multiply our accumulator from step 1 by 10 and add to it the digit that we got in step 2.
+4. Move to the next symbol in the given string and repeat steps 2 and 3 if it is not the end of the string (`\0` symbol).
+
+Returning to the table from the section above, we may see that pointers to the command line arguments are located on the stack right above the number of command line arguments. So if we fetch the first value from the stack after we already fetched the number of arguments, it will be a pointer to the string which is the first command line argument. If we will pop the next value from the stack, it will be the second command line argument passed to the program.
```

Comment:
```suggestion
Returning to the table from the section above, we can see that pointers to the command-line arguments are located on the stack right above the number of command-line arguments. So, after we pop the number of arguments (ARGC), the stack pointer will point to the address of the first command-line argument (ARGV[0]). If we pop the next value from the stack, it will point to the second command-line argument (ARGV[1]) passed to the program.
```